automation/freebsd_repo/FreeBSD_9.2_Repo/src/usr.bin/uniq/uniq.1
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     From: @(#)uniq.1	8.1 (Berkeley) 6/6/93
.\" $FreeBSD: releng/9.2/usr.bin/uniq/uniq.1 216370 2010-12-11 08:32:16Z joel $
.\"
.Dd December 17, 2009
.Dt UNIQ 1
.Os
.Sh NAME
.Nm uniq
.Nd report or filter out repeated lines in a file
.Sh SYNOPSIS
.Nm
.Op Fl c | Fl d | Fl u
.Op Fl i
.Op Fl f Ar num
.Op Fl s Ar chars
.Oo
.Ar input_file
.Op Ar output_file
.Oc
.Sh DESCRIPTION
The
.Nm
utility reads the specified
.Ar input_file
comparing adjacent lines, and writes a copy of each unique input line to
the
.Ar output_file .
If
.Ar input_file
is a single dash
.Pq Sq Fl
or absent, the standard input is read.
If
.Ar output_file
is absent, standard output is used for output.
The second and succeeding copies of identical adjacent input lines are
not written.
Repeated lines in the input will not be detected if they are not adjacent,
so it may be necessary to sort the files first.
.Pp
The following options are available:
.Bl -tag -width Ds
.It Fl c
Precede each output line with the count of the number of times the line
occurred in the input, followed by a single space.
.It Fl d
Only output lines that are repeated in the input.
.It Fl f Ar num
Ignore the first
.Ar num
fields in each input line when doing comparisons.
A field is a string of non-blank characters separated from adjacent fields
by blanks.
Field numbers are one based, i.e., the first field is field one.
.It Fl s Ar chars
Ignore the first
.Ar chars
characters in each input line when doing comparisons.
If specified in conjunction with the
.Fl f
option, the first
.Ar chars
characters after the first
.Ar num
fields will be ignored.
Character numbers are one based, i.e., the first character is character one.
.It Fl u
Only output lines that are not repeated in the input.
.It Fl i
Case insensitive comparison of lines.
.\".It Fl Ns Ar n
.\"(Deprecated; replaced by
.\".Fl f ) .
.\"Ignore the first n
.\"fields on each input line when doing comparisons,
.\"where n is a number.
.\"A field is a string of non-blank
.\"characters separated from adjacent fields
.\"by blanks.
.\".It Cm \&\(pl Ns Ar n
.\"(Deprecated; replaced by
.\".Fl s ) .
.\"Ignore the first
.\".Ar m
.\"characters when doing comparisons, where
.\".Ar m
.\"is a
.\"number.
.El
.Sh ENVIRONMENT
The
.Ev LANG ,
.Ev LC_ALL ,
.Ev LC_COLLATE
and
.Ev LC_CTYPE
environment variables affect the execution of
.Nm
as described in
.Xr environ 7 .
.Sh EXIT STATUS
.Ex -std
.Sh COMPATIBILITY
The historic
.Cm \&\(pl Ns Ar number
and
.Fl Ns Ar number
options have been deprecated but are still supported in this implementation.
.Sh SEE ALSO
.Xr sort 1
.Sh STANDARDS
The
.Nm
utility conforms to
.St -p1003.1-2001
as amended by Cor.\& 1-2002.
.Sh HISTORY
A
.Nm
command appeared in
.At v3 .


automation/freebsd_repo/FreeBSD_9.2_Repo/src/usr.sbin/services_mkdb/uniq.c
/*	$NetBSD: uniq.c,v 1.4 2008/04/28 20:24:17 martin Exp $	*/

/*-
 * Copyright (c) 2007 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Christos Zoulas.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <sys/cdefs.h>
__FBSDID("$FreeBSD: releng/9.2/usr.sbin/services_mkdb/uniq.c 206156 2010-04-04 08:38:14Z ume $");

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <db.h>
#include <err.h>
#include <libutil.h>
#include <ctype.h>
#include <fcntl.h>

extern const HASHINFO hinfo;

void uniq(const char *);
static int comp(const char *, char **, size_t *);

/*
 * Preserve only unique content lines in a file. Input lines that have
 * content [alphanumeric characters before a comment] are white-space
 * normalized and have their comments removed. Then they are placed
 * in a hash table, and only the first instance of them is printed.
 * Comment lines without any alphanumeric content are always printed
 * since they are there to make the file "pretty". Comment lines with
 * alphanumeric content are also placed into the hash table and only
 * printed once.
 */
void
uniq(const char *fname)
{
	DB *db;
	DBT key;
	static const DBT data = { NULL, 0 };
	FILE *fp;
	char *line;
	size_t len;

	if ((db = dbopen(NULL, O_RDWR, 0, DB_HASH, &hinfo)) == NULL)
		err(1, "Cannot create in memory database");

	if ((fp = fopen(fname, "r")) == NULL)
		err(1, "Cannot open `%s'", fname);
	while ((line = fgetln(fp, &len)) != NULL) {
		size_t complen = len;
		char *compline;
		if (!comp(line, &compline, &complen)) {
			(void)fprintf(stdout, "%*.*s", (int)len, (int)len,
			    line);
			continue;
		}
		key.data = compline;
		key.size = complen;
		switch ((db->put)(db, &key, &data, R_NOOVERWRITE)) {
		case 0:
			(void)fprintf(stdout, "%*.*s", (int)len, (int)len,
			    line);
			break;
		case 1:
			break;
		case -1:
			err(1, "put");
		default:
			abort();
			break;
		}
	}
	(void)fflush(stdout);
	exit(0);
}

/*
 * normalize whitespace in the original line and place a new string
 * with whitespace converted to a single space in compline. If the line
 * contains just comments, we preserve them. If it contains data and
 * comments, we kill the comments. Return 1 if the line had actual
 * contents, or 0 if it was just a comment without alphanumeric characters.
 */
static int
comp(const char *origline, char **compline, size_t *len)
{
	const unsigned char *p;
	unsigned char *q;
	char *cline;
	size_t l = *len, complen;
	int hasalnum, iscomment;

	/* Eat leading space */
	for (p = (const unsigned char *)origline; l && *p && isspace(*p);
	    p++, l--)
		continue;
	if ((cline = malloc(l + 1)) == NULL)
		err(1, "Cannot allocate %zu bytes", l + 1);
	(void)memcpy(cline, p, l);
	cline[l] = '\0';
	if (*cline == '\0')
		return 0;

	complen = 0;
	hasalnum = 0;
	iscomment = 0;

	for (q = (unsigned char *)cline; l && *p; p++, l--) {
		if (isspace(*p)) {
			if (complen && isspace(q[-1]))
				continue;
			*q++ = ' ';
			complen++;
		} else {
			if (!iscomment && *p == '#') {
				if (hasalnum)
					break;
				iscomment = 1;
			} else
				hasalnum |= isalnum(*p);
			*q++ = *p;
			complen++;
		}
	}

	/* Eat trailing space */
	while (complen && isspace(q[-1])) {
		--q;
		--complen;
	}
	*q = '\0';
	*compline = cline;
	*len = complen;
	return hasalnum;
}
